Improve KOTH by adding donation events

sh-dima opened the pull request:
Improved the King of the Hill minigame by adding donation events.

sh-dima committed:
feat: add basic delayed kb to king of the hill

Co-authored-by: rolyPolyVole <yezhanting@gmail.com>

sh-dima committed:
fix: correct calculating delayed kb ticks

sh-dima committed:
fix: check if delayed kb is not active

sh-dima committed:
fix: reduce delayed kb

sh-dima committed:
fix: calculate current index correctly

sh-dima committed:
debug: add logs

sh-dima committed:
fix: move logic around

sh-dima committed:
fix: throw around players for just one more tick to fix a bug

sh-dima committed:
fix: fix logic errors for calculations

sh-dima committed:
debug: log end of KB period

sh-dima committed:
feat: actually apply no KB in delayed KB period

sh-dima committed:
fix: actually calculate index

sh-dima committed:
chore: remove logs

sh-dima committed:
fix: change strategy to calculate index

sh-dima committed:
fix: don't constantly clear velocity map

sh-dima committed:
fix: don't reset ticks left on donation

sh-dima committed:
fix: add velocity onto the players velocity

sh-dima committed:
chore: remove logs

sh-dima committed:
fix: fix potential exceptions

sh-dima committed:
feat: scale velocity according to player's item

sh-dima committed:
feat: make throw around period shorter

sh-dima committed:
test: try even faster throw around period

sh-dima committed:
test: try resetting player velocity

sh-dima committed:
feat: nerf KB

rolyPolyVole committed:
ideas: add KOTH ideas

sh-dima committed:
feat: add progress bossbar

Co-authored-by: rolyPolyVole <yezhanting@gmail.com>

sh-dima committed:
feat: add double jump logic

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: add missing flight allowance

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add double jump event

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add double jump action bar

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add position shuffling event

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add slow falling event

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: handle multiple low tier donations

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add 3-5 double jumps

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add KING'S BLINDNESS event

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: add KING'S BLINDNESS to low tier events

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: stop players from flying

sh-dima committed:
fix: check for correct gamemode

sh-dima committed:
feat: make sure positions are shuffled

sh-dima committed:
fix: run other event if blindness fails

sh-dima committed:
fix: align shuffling with countdown

sh-dima committed:
fix: avoid infinite loop

sh-dima committed:
fix: compare by equality, not reference

sh-dima committed:
fix: prevent division by zero

sh-dima committed:
chore: remove unused imports

sh-dima committed:
fix: don't delay shuffling positions

sh-dima committed:
feat: add jump boost event

sh-dima committed:
feat: increase KOTH time

90 seconds really isn't enough to be interesting.

sh-dima committed:
feat: add more void to KOTH

This gives players a better chance to escape using their double jumps.

sh-dima committed:
fix: don't use donation event sound

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: play button click sound in countdown

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
feat: play note block ping sound in countdown

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: prevent players from permanently having KB resistance

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
chore: handle multiple medium tier donations

sh-dima committed:
feat: add invisibility event

rolyPolyVole committed:
chore: add more ideas to koth

sh-dima committed:
feat: increase jump boost amplifier

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: fix invisibility text

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: fix stick attribute

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: actually update item meta

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: play high-pitched sound 1 tick later

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: make sure that effects can stack duration

sh-dima committed:
fix: fix attribute bug

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: actually make players invisible

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
fix: use correct time unit

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

sh-dima committed:
Revert "fix: use correct time unit"

This reverts commit 4d825be5c2b463408d91d2062ae7c4d61f1ac7fc.

sh-dima committed:
Revert "fix: actually make players invisible"

This reverts commit 12731100b296b6d35ea77c1812783b2624200084.

sh-dima committed:
feat: play wither block break sound when thrown around

Co-authored-by: rolyPolyVole <108834596+rolypolyvole@users.noreply.github.com>

ShreyasAyyengar committed:
Merge branch 'master' into kinghill-donation-events

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/minigame/games/KingHill.kt:
```diff
@@ -13,30 +13,48 @@ import gg.flyte.twilight.extension.playSound
-    private var gameTime = 90
-    private val respawnBelow = 71
+    private var gameTime = 150
+    private val respawnBelow = 60
```
Any justification for this change?

sh-dima replied:
Yeah, I believe in my commit messages I said the following:
- I decreased the `respawnBelow` value as this gives double jumps more use (save yourself from falling into the void, the previous height was too high to react fast enough).
- I increased the game time as 90 seconds felt really short during my testing with @rolyPolyVole, and with even more players I don't think it will be long enough to have an interesting game. You can change it back if you like but I think this is the correct decision.

ShreyasAyyengar committed:
refactor: change variables and add variables

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
refactor: simplify method

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
refactor: move methods around to nest

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
style: add clarifying paren

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
docs/style/refactor: add clarifying docs and paren, reduce whitespace

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
refactor: use more coherent method

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
refactor: rename

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
refactor: restructured donation events.

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/minigame/games/KingHill.kt:
```diff
@@ -91,9 +112,47 @@ class KingHill : EventMiniGame(GameConfig.KING_OF_THE_HILL) {
 
                 updateScoreboard()
             }
+
+            tasks += repeatingTask(1) {
+                val (ticksLeft, totalTicks) = thrownAroundTickData
+
+                if (ticksLeft == 0) {
+                    velocityMap.clear()
+                    thrownAroundTickData = 0 to 0
+                    return@repeatingTask
+                }
+
+                velocityMap.entries.forEach {
+                    val player = Bukkit.getPlayer(it.key)!!
+                    val vectors = it.value
+                    val vectorCount = vectors.size
+
+                    if (vectorCount == 0) return@forEach
+
+                    val ticksPassed = totalTicks - ticksLeft
+                    val previousTicksPassed = totalTicks - (ticksLeft + 1)
+
+                    val ratio = ticksPassed.toDouble() / totalTicks.toDouble()
+                    val previousRatio = previousTicksPassed.toDouble() / totalTicks.toDouble()
+
+                    val index = ratio * vectorCount
+                    val previousIndex = previousRatio * vectorCount
+
+                    val floor = floor(index)
+                    val previousFloor = floor(previousIndex)
+
+                    if (floor == previousFloor) return@forEach
+                    player.velocity = vectors[floor.toInt()]
+                }
+
+                thrownAroundTickData = ticksLeft - 1 to totalTicks
+            }
+
```
I'd like to see some documentation or comments on what this does exactly.

ShreyasAyyengar committed:
refactor: suppress at lower level

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar committed:
refactor: remove unnecessary param

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

sh-dima committed:
docs/refactor: make throwing around feature code clearer

sh-dima committed:
fix/refactor: don't create attribute modifier unnecessarily

sh-dima committed:
fix: add back very much necessary param

Without this parameter, the countdown starts 1 second late, and so it appears as if players are swapped 1 second early (in other words, the exact moment when the countdown says "Players will be swapped in 1 seconds").

This reverts commit 184e65e49de4e12a6a690f2b960c0fba9d3b6022.

sh-dima committed:
refactor: add explicit parameter

sh-dima committed:
refactor: add explicit parameter

sh-dima committed:
fix: fix constant wither break block sound

sh-dima committed:
fix: fix possible NPE

sh-dima committed:
debug: add logs to fix delayed kb

sh-dima committed:
fix: prevent constant spam

sh-dima committed:
fix: prevent high negative data

sh-dima committed:
fix: attempt to fix delayed kb

sh-dima committed:
debug: add more logs

sh-dima committed:
debug: add more logs

sh-dima committed:
debug: add more logs

sh-dima committed:
debug: add more logs

sh-dima committed:
fix: set total ticks to -1 also

sh-dima committed:
fix: use -1

sh-dima committed:
debug: add more logs

sh-dima committed:
debug: amend log

sh-dima committed:
fix: fix throwing around period

sh-dima committed:
fix: make sure invisibility effect stacks

sh-dima committed:
chore: remove logs

sh-dima commented:
I believe everything is fixed. I tested with my alt account just now and it seems to work.

ShreyasAyyengar commented:
[Thanks](https://github.com/flytegg/christmas-plugin-2024/commit/ff91f16cae4fdbf70842509e4c20af89e501c20d)

Do send or PM me a Minecraft UUID so I can add you as an NPC in our contributors section as a thank you for your contribution. Send it along with a HEX colour code `#abc123` with the preferred colour of your NPC nameplate.
